File: app/src/lib/contract.ts
/**
 * Wire contract for the DA3 inference service.
 *
 * Mirrors `server/contract.py` field for field — change both together.
 * Defaults are the verified upstream values; see docs/SOURCES.md.
 *
 * ONE DELIBERATE DIVERGENCE (2026-08-04): the server still declares `infer_gs: bool = False`
 * and the `gs_ply` / `gs_video` artifact kinds. Gaussian splats were dropped from the roadmap
 * — no measurement node consumes them — so the client no longer sends or understands them.
 * The server side is left alone on purpose: editing `server/` changes the image source hash
 * and forces a ~20 min rebuild for dead code. It has a default, so omitting the field is safe.
 */

export const SCHEMA_VERSION = "verge.infer-manifest/0.1.0";

export const MODEL_REPOSITORY_ID = "depth-anything/DA3NESTED-GIANT-LARGE-1.1";
export const MODEL_REVISION = "b2359bdf726fb44ef62acca04d629dcf158053e7";

/**
 * The L4 we deploy on — the denominator for the VRAM budget bar.
 * MEASURED from `torch.cuda.mem_get_info()` on the real device: 22.03 GiB usable.
 * An L4's advertised "24 GB" is decimal and some is reserved, so 24 GiB was wrong.
 */
export const L4_TOTAL_VRAM_BYTES = 23_659_151_360;

/** VRAM resident after the model loads, before any inference. Measured. */
export const MODEL_RESIDENT_BYTES = 7_050_625_024;

/**
 * Frame cap, set from the 2026-08-01 ladder on a real L4.
 *
 * MEASURED CEILING: 144 frames @ 504 px ran (21.88 GiB of 22.03); 160 OOMed. So the
 * hard ceiling sits between 144 and 160.
 *
 * 112 is deliberately BELOW that, not at it. Both 128 and 144 completed at ~99% of the
 * device, which is not an operating point — any variation in scene content or allocator
 * state crosses the line. 112 measured 17.23 GiB of allocator against a ~19.5 GiB
 * ceiling, roughly 15% headroom, and matched the fitted model to within 0.25 GiB.
 *
 * On the 26.61 s test clip this is 4.21 effective fps, up from 1.20 at the old cap of 32.
 */
export const DEFAULT_MAX_FRAMES = 112;

/**
 * Measured peak VRAM from scripts/vram-sweep.sh on a real L4 (2026-08-01), at 504 px.
 * Raw data, including the 356/252 px ladders, in docs/vram-measurements.json.
 *
 * These are DRIVER peaks with each run isolated by empty_cache() +
 * reset_peak_memory_stats(), so unlike the 2026-07-31 set they are per-run costs rather
 * than cumulative high-water marks. The driver figure includes the CUDA context and
 * allocator reserve, so it saturates near the device limit; the cleaner signal for
 * modelling is the allocator peak, which fits
 *
 *     allocator ≈ 0.0700 GiB/frame + 9.39 GiB
 *
 * across 32/64/128/144 frames. Note that is HALF the 0.14 GiB/frame the old
 * contaminated data implied — which is why the real ceiling landed at ~144 rather
 * than the predicted ~110.
 */
const GIB = 1_073_741_824;

/** The rungs exactly as measured, in frame order. */
const MEASURED_DRIVER_PEAKS: ReadonlyArray<{ frames: number; peakBytes: number }> = [
  { frames: 32, peakBytes: 15_294_529_536 },
  { frames: 64, peakBytes: 18_186_502_144 },
  { frames: 112, peakBytes: 22_848_471_040 },
  { frames: 128, peakBytes: 23_561_502_720 },
  // The 144 rung was overwritten by the sweep-merge bug and recovered from console output, so
  // only its GiB figure survives (`"recovered": true`, null peak_bytes in the JSON). Derived
  // here rather than back-computed to fake byte precision that was genuinely lost.
  { frames: 144, peakBytes: Math.round(21.88 * GIB) },
];

/**
 * The table the UI interpolates, forced non-decreasing.
 *
 * 144 measures LOWER than 128 (21.88 vs 21.94 GiB). VRAM did not fall: the driver figure
 * includes the CUDA context and allocator reserve, so past ~21.5 GiB it is a plateau against
 * the 22.03 GiB device limit with measurement scatter on top, not a curve. The allocator series
 * is the one that stays clean and monotonic (18.35 → 19.47 GiB across the same rungs). Taking a
 * running maximum keeps the estimate physical without editing the measurements above.
 */
export const VRAM_MEASUREMENTS: ReadonlyArray<{ frames: number; peakBytes: number }> =
  MEASURED_DRIVER_PEAKS.map((point, index, all) => ({
    frames: point.frames,
    peakBytes: Math.max(...all.slice(0, index + 1).map((entry) => entry.peakBytes)),
  }));

/** Highest frame count we have actually run to success. Beyond this we are extrapolating. */
export const MAX_MEASURED_FRAMES = 144;

/** Lowest frame count measured to OOM at 504 px. The ceiling is in (144, 160). */
export const MIN_OOM_FRAMES = 160;
/** The resolution every measurement was taken at. */
export const MEASURED_PROCESS_RES = 504;

export interface VramPrediction {
  bytes: number;
  /** True only when interpolating inside measured data at the measured resolution. */
  measured: boolean;
}

/**
 * Peak VRAM for a given frame count, interpolated from real measurements.
 * Anything outside the measured envelope is flagged so the UI can say so.
 */
export function predictVram(frameCount: number, processRes: number): VramPrediction {
  const inResRange = processRes === MEASURED_PROCESS_RES;
  const points = VRAM_MEASUREMENTS;
  const first = points[0]!;
  const last = points[points.length - 1]!;

  let bytes: number;
  // `measured` means this exact configuration was actually run — interpolating
  // between two real points is still a guess, even when both ends agree.
  const exact = points.some((p) => p.frames === frameCount);

  if (frameCount <= first.frames) {
    // Below the measured floor: scale down from the first point, but never below
    // what the model alone occupies.
    bytes = Math.max(MODEL_RESIDENT_BYTES, first.peakBytes * (frameCount / first.frames));
  } else if (frameCount >= last.frames) {
    // Extrapolate on the slope of the last measured segment.
    const prev = points[points.length - 3]!; // last distinct step
    const slope = (last.peakBytes - prev.peakBytes) / (last.frames - prev.frames);
    bytes = last.peakBytes + slope * (frameCount - last.frames);
  } else {
    let lo = first;
    let hi = last;
    for (let i = 0; i < points.length - 1; i++) {
      if (frameCount >= points[i]!.frames && frameCount <= points[i + 1]!.frames) {
        lo = points[i]!;
        hi = points[i + 1]!;
        break;
      }
    }
    const span = hi.frames - lo.frames;
    const t = span === 0 ? 0 : (frameCount - lo.frames) / span;
    bytes = lo.peakBytes + t * (hi.peakBytes - lo.peakBytes);
  }

  // Resolution scales the token count quadratically; only 504 was measured.
  const resScale = (processRes / MEASURED_PROCESS_RES) ** 2;
  return { bytes: bytes * resScale, measured: exact && inResRange };
}

export type ProcessResMethod = "upper_bound_resize" | "lower_bound_resize";
export type RefViewStrategy = "first" | "middle" | "saddle_balanced" | "saddle_sim_range";
/**
 * "npz" is the bundle the service writes with the key names `npz.ts` reads.
 * "npz_native" is DA3's own export — kept and served, but distinguished so a
 * lookup by kind can never pick it up by accident.
 */
export type ArtifactKind = "glb" | "npz" | "npz_native" | "depth_preview";

export interface InferParams {
  /** Sampling rate the local ffmpeg pass used. The server never resamples. */
  fps: number;
  sourceDurationS?: number;
  processRes: number;
  processResMethod: ProcessResMethod;
  refViewStrategy: RefViewStrategy;
  maxFrames: number;
}

/** Verified upstream defaults. The HF Space's own sampling slider defaults to 10 fps. */
export const DEFAULT_PARAMS: InferParams = {
  fps: 10,
  processRes: 504,
  processResMethod: "upper_bound_resize",
  refViewStrategy: "middle",
  maxFrames: DEFAULT_MAX_FRAMES,
};

/** UI bounds for the fps control. DA3's own slider allows 0.1-60; we start at 1. */
export const FPS_MIN = 1;
export const FPS_MAX = 50;

export interface Artifact {
  kind: ArtifactKind;
  name: string;
  sizeBytes: number;
  sha256: string;
  url: string;
}

/**
 * Two peaks, because they answer different questions and routinely disagree.
 * `peakBytes` is the driver high-water mark — use it to decide whether a
 * configuration fits. `torchPeakBytes` is the caching allocator's, isolated to this
 * run by an `empty_cache()` at entry — use it, with `baselineBytes`, to model what an
 * extra frame costs. See server/vram.py for why the driver number alone was misleading.
 *
 * Both new fields are 0 on manifests written before 2026-08-01.
 */
export interface VramStats {
  peakBytes: number;
  currentBytes: number;
  totalBytes: number;
  deviceName: string;
  torchPeakBytes: number;
  baselineBytes: number;
}

/** Observations recorded during a run so open questions need no extra cloud session. */
export interface Diagnostics {
  /** Each npz DA3 itself wrote, mapped to its array key names. */
  nativeNpz: Record<string, string[]>;
  exportDirListing: string[];
}

export interface Timing {
  gpuSeconds: number;
  wallSeconds: number;
  modelLoadSeconds?: number;
}

export interface FrameInfo {
  count: number;
  requestedCount: number;
  width: number;
  height: number;
  /** True when fps x duration exceeded maxFrames and we sampled down. */
  capped: boolean;
  effectiveFps?: number;
}

export interface InferManifest {
  schemaVersion: string;
  runId: string;
  modelRepositoryId: string;
  modelRevision: string;
  depthMode: "metric";
  linearUnit: "metre";
  params: InferParams;
  frames: FrameInfo;
  timing: Timing;
  vram: VramStats;
  artifacts: Artifact[];
  diagnostics?: Diagnostics;
  transient: boolean;
  expiresAfterDays: number;
  /**
   * True when this manifest came from the fixture-backed dev mock rather than a GPU.
   * The real service never sets it. Surfaced on the node card because a screenshot of
   * a mock run is otherwise indistinguishable from a real one — see docs/DESIGN.md's
   * honesty rules.
   */
  mock?: boolean;
}

export interface GpuSnapshot {
  available: boolean;
  modelLoaded: boolean;
  busy: boolean;
  deviceName: string;
  currentBytes: number;
  peakBytes: number;
  totalBytes: number;
}

export type GpuState = "cold" | "warming" | "warm" | "busy" | "error";

export function vramFraction(bytes: number, total: number): number {
  return total > 0 ? bytes / total : 0;
}

export function formatBytes(bytes: number): string {
  if (bytes <= 0) return "0 B";
  const gib = bytes / 1024 ** 3;
  if (gib >= 1) return `${gib.toFixed(2)} GiB`;
  const mib = bytes / 1024 ** 2;
  return `${mib.toFixed(0)} MiB`;
}

/**
 * How many frames an fps/duration pair yields, and what actually gets sent once
 * the cap applies. Mirrors DA3's `frame_interval = max(1, int(video_fps / fps))`
 * behaviour, then clamps. Never silently truncates the clip — it lowers fps
 * instead, so the frames still span the whole video.
 */
export function planFrames(
  fps: number,
  durationS: number,
  maxFrames: number,
): { count: number; effectiveFps: number; capped: boolean } {
  const requested = Math.max(1, Math.floor(fps * durationS));
  if (requested <= maxFrames) {
    return { count: requested, effectiveFps: fps, capped: false };
  }
  return {
    count: maxFrames,
    effectiveFps: durationS > 0 ? maxFrames / durationS : fps,
    capped: true,
  };
}

File: server/contract.py
"""Wire contract for the DA3 inference service.

This module is the single source of truth for the /infer request and response
shapes. `app/src/lib/contract.ts` mirrors it field for field — change both together.

Defaults here are the verified upstream values; see docs/SOURCES.md.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

SCHEMA_VERSION = "verge.infer-manifest/0.1.0"

MODEL_REPOSITORY_ID = "depth-anything/DA3NESTED-GIANT-LARGE-1.1"
MODEL_REVISION = "b2359bdf726fb44ef62acca04d629dcf158053e7"

# MEASURED from torch.cuda.mem_get_info() on the deployed L4 (2026-07-31): 22.03 GiB
# usable. The advertised "24 GB" is decimal and some is reserved.
L4_TOTAL_VRAM_BYTES = 23_659_151_360

# Frame-count ceiling — a safety rail, not a quality target (DA3 imposes none).
# Set from the 2026-08-01 ladder in docs/vram-measurements.json. MEASURED: 144 frames
# @ 504 px ran at 21.88 GiB of 22.03; 160 OOMed. 112 sits deliberately below that
# ceiling with ~15% headroom (17.23 GiB allocator), because 128 and 144 both completed
# at ~99% of the device, which is a coin flip rather than an operating point.
# Mirrors DEFAULT_MAX_FRAMES in app/src/lib/contract.ts.
DEFAULT_MAX_FRAMES = 112

ProcessResMethod = Literal["upper_bound_resize", "lower_bound_resize"]
RefViewStrategy = Literal["first", "middle", "saddle_balanced", "saddle_sim_range"]
ExportFormat = Literal["npz", "mini_npz", "glb", "gs_ply", "gs_video"]


class InferParams(BaseModel):
    """Inference knobs. Every one of these is surfaced in the app's inspector so
    settings can be compared across runs."""

    # Informational: the sampling rate the client's ffmpeg pass actually used.
    # The server does not resample; frames arrive already extracted (local-first).
    fps: float = Field(default=10.0, gt=0, le=60)
    source_duration_s: float | None = Field(default=None, gt=0)

    process_res: int = Field(default=504, ge=126, le=1024)
    process_res_method: ProcessResMethod = "upper_bound_resize"
    ref_view_strategy: RefViewStrategy = "middle"
    infer_gs: bool = False

    # Upper bound raised 200 -> 512 on 2026-08-01: the frame ladder runs to 256, and at
    # 200 pydantic rejected the request before it could reach the GPU and OOM, which is
    # the datapoint the ladder exists to collect. This bound is a wire sanity check, not
    # the safety rail — that is max_frames itself, defaulting to DEFAULT_MAX_FRAMES.
    max_frames: int = Field(default=DEFAULT_MAX_FRAMES, ge=2, le=512)


class Artifact(BaseModel):
    # "npz" is OUR array bundle, with the key names app/src/lib/npz.ts reads.
    # "npz_native" is whatever DA3's own exporter wrote — kept, served, but never
    # mistaken for ours by a client looking artifacts up by kind.
    kind: Literal["glb", "npz", "npz_native", "depth_preview", "gs_ply", "gs_video"]
    name: str
    size_bytes: int
    sha256: str
    # Either a signed GCS URL or a path on this service (/artifact/{run_id}/{name}).
    url: str


class VramStats(BaseModel):
    """Peak is what matters for capacity planning; current is what the live
    telemetry poll reports mid-run.

    `peak_bytes` (driver) and `torch_peak_bytes` (caching allocator) answer different
    questions and routinely disagree — see the module docstring in server/vram.py.
    Use the driver peak to decide whether a configuration fits; use the allocator peak
    and `baseline_bytes` to model what an extra frame costs.

    The two new fields default to 0 so manifests written before 2026-08-01 still parse.
    """

    peak_bytes: int
    current_bytes: int
    total_bytes: int
    device_name: str
    # Allocator high-water mark, isolated to this run by empty_cache() at entry.
    torch_peak_bytes: int = 0
    # Driver usage at run start, after the cache was dropped: model + CUDA context.
    baseline_bytes: int = 0

    @property
    def peak_fraction(self) -> float:
        return self.peak_bytes / self.total_bytes if self.total_bytes else 0.0

    @property
    def activation_bytes(self) -> int:
        """Marginal cost of this run's frames, above the resident model."""
        return max(0, self.peak_bytes - self.baseline_bytes)


class Timing(BaseModel):
    gpu_seconds: float
    wall_seconds: float
    model_load_seconds: float | None = None


class FrameInfo(BaseModel):
    count: int
    requested_count: int
    width: int
    height: int
    # True when fps x duration exceeded max_frames and we sampled down.
    capped: bool = False
    effective_fps: float | None = None


class Diagnostics(BaseModel):
    """Facts observed during a run that answer open questions without spending
    another cloud session on them. Purely additive — nothing downstream requires it.

    `native_npz` maps each npz DA3 wrote to its key names, which is how we confirm
    what DA3 calls its own arrays rather than guessing from the fixture.
    """

    native_npz: dict[str, list[str]] = Field(default_factory=dict)
    export_dir_listing: list[str] = Field(default_factory=list)


class InferManifest(BaseModel):
    schema_version: str = SCHEMA_VERSION
    run_id: str
    model_repository_id: str = MODEL_REPOSITORY_ID
    model_revision: str = MODEL_REVISION
    depth_mode: Literal["metric"] = "metric"
    linear_unit: Literal["metre"] = "metre"

    params: InferParams
    frames: FrameInfo
    timing: Timing
    vram: VramStats
    artifacts: list[Artifact]
    diagnostics: Diagnostics | None = None

    # Transient by policy — the client must call Save explicitly to retain.
    transient: bool = True
    expires_after_days: int = 3


class GpuSnapshot(BaseModel):
    """Polled by the UI during inference for the live VRAM readout."""

    available: bool
    model_loaded: bool
    busy: bool
    device_name: str = ""
    current_bytes: int = 0
    peak_bytes: int = 0
    total_bytes: int = L4_TOTAL_VRAM_BYTES


class HealthResponse(BaseModel):
    status: Literal["ok"]
    model_loaded: bool
    gpu_available: bool
    model_revision: str = MODEL_REVISION


class WarmupResponse(BaseModel):
    model_loaded: bool
    model_load_seconds: float
    gpu: GpuSnapshot
